File: core/src/main/java/xtdb/bitemporal/Polygon.java
package xtdb.bitemporal;

import com.carrotsearch.hppc.LongArrayList;

public record Polygon(LongArrayList validTimes, LongArrayList sysTimeCeilings) implements IPolygonReader {
    public Polygon() {
        this(new LongArrayList(), new LongArrayList());
    }

    @Override
    public int getValidTimeRangeCount() {
        return sysTimeCeilings.elementsCount;
    }

    @Override
    public long getValidFrom(int rangeIdx) {
        return validTimes.get(rangeIdx);
    }

    @Override
    public long getValidTo(int rangeIdx) {
        return validTimes.get(rangeIdx + 1);
    }

    @Override
    public long getSystemTo(int rangeIdx) {
        return sysTimeCeilings.get(rangeIdx);
    }

    public void calculateFor(Ceiling ceiling, IPolygonReader inPolygon) {
        validTimes.clear();
        sysTimeCeilings.clear();

        var ceilIdx = 0;

        var validTimeRangeCount = inPolygon.getValidTimeRangeCount();
        assert validTimeRangeCount > 0;

        int rangeIdx = 0;

        var validTime = inPolygon.getValidFrom(rangeIdx);
        var validTo = inPolygon.getValidTo(rangeIdx);
        var systemTo = inPolygon.getSystemTo(rangeIdx);

        while (true) {
            if (systemTo != Long.MAX_VALUE) {
                validTimes.add(validTime);
                sysTimeCeilings.add(systemTo);
                validTime = validTo;
            } else {
                long ceilValidTo;

                while(true) {
                    ceilValidTo = ceiling.getValidTo(ceilIdx);
                    if (ceilValidTo > validTime) break;
                    ceilIdx++;
                }

                validTimes.add(validTime);
                sysTimeCeilings.add(ceiling.getSystemTime(ceilIdx));

                validTime = Math.min(ceilValidTo, validTo);
            }

            if (validTime == validTo) {
                if (++rangeIdx == validTimeRangeCount) break;

                validTo = inPolygon.getValidTo(rangeIdx);
                systemTo = inPolygon.getSystemTo(rangeIdx);
            }
        }

        validTimes.add(validTime);
    }
}

File: core/src/main/kotlin/xtdb/bitemporal/Polygon.kt
package xtdb.bitemporal

import com.carrotsearch.hppc.LongArrayList
import xtdb.trie.EventRowPointer
import kotlin.math.min
import kotlin.math.max

data class Polygon(val validTimes: LongArrayList, val sysTimeCeilings: LongArrayList) {
    constructor() : this(LongArrayList(), LongArrayList())

    val validTimeRangeCount
        get() = sysTimeCeilings.elementsCount

    fun getValidFrom(rangeIdx: Int) = validTimes[rangeIdx]

    fun getValidTo(rangeIdx: Int) = validTimes[rangeIdx + 1]

    fun getSystemTo(rangeIdx: Int) = sysTimeCeilings[rangeIdx]

    /**
     * maximum value T where row is invalid for all (valid-time, sys-time) where both valid-time >= T and sys-time >= T
     */
    val recency: Long
        get() {
            val validTimeRangeCount = validTimeRangeCount

            // largest recency we've seen so far
            var recency = Long.MIN_VALUE

            var validTo = getValidTo(validTimeRangeCount - 1)

            // start from the RHS
            for (i in (validTimeRangeCount - 1) downTo 0) {
                recency = max(recency, min(getSystemTo(i), validTo))

                val validFrom = getValidFrom(i)

                // it's never going to get bigger than this, stop early
                if (recency >= validFrom) return recency

                validTo = validFrom
            }

            return recency
        }

    fun calculateFor(ceiling: Ceiling, validFrom: Long, validTo: Long) {
        validTimes.clear()
        sysTimeCeilings.clear()

        var validTime = validFrom

        var ceilIdx = ceiling.getCeilingIndex(validFrom)

        do {
            var ceilValidTo: Long

            while (true) {
                ceilValidTo = ceiling.getValidTo(ceilIdx)
                if (ceilValidTo > validTime) break
                ceilIdx++
            }

            validTimes.add(validTime)
            sysTimeCeilings.add(ceiling.getSystemTime(ceilIdx))

            validTime = min(ceilValidTo, validTo)
        } while (validTime != validTo)

        validTimes.add(validTime)
    }

    fun calculateFor(ceiling: Ceiling, evPtr: EventRowPointer) {
        calculateFor(ceiling, evPtr.validFrom, evPtr.validTo)
    }
}
